datafusion/core/src/physical_optimizer/repartition.rs

```diff
@@ -163,10 +166,14 @@ fn optimize_partitions(
             .children()
             .iter()
             .map(|child| {
+                let can_reorder_child = (can_reorder
+                    && plan.required_input_ordering().is_empty())
```

Comment:
The change is to add `&& plan.required_input_ordering().is_empty())` to the check if the child can be reordered